Objective-C/LogSink/CBLLogSinks.mm

```diff
@@ -0,0 +1,211 @@
+    CBL_LOCK(self) {
+        [self checkLogApiVersion: LogAPINew];
+        _console = console;
+    }
+    [self updateLogLevels];
+}
+
++ (CBLConsoleLogSink*) console {
+    CBL_LOCK(self) {
+        [self checkLogApiVersion: LogAPINew];
+        return _console;
+    }
+}
+
++ (void) setCustom: (CBLCustomLogSink*) custom {
+    CBL_LOCK(self) {
+        [self checkLogApiVersion: LogAPINew];
+        _custom = custom;
+    }
+    [self updateLogLevels];
+}
+
++ (CBLCustomLogSink*) custom {
+    CBL_LOCK(self) {
+        [self checkLogApiVersion: LogAPINew];
+        return _custom;
+    }
+}
+
++ (void) setFile: (CBLFileLogSink*) file {
+    CBL_LOCK(self) {
+        [self checkLogApiVersion: LogAPINew];
+    }
+    
+    _file = file;
+    [CBLFileLogSink setup: file];
+    [self updateLogLevels];
+}
+
++ (CBLFileLogSink*) file {
+    CBL_LOCK(self) {
+        [self checkLogApiVersion: LogAPINew];
+        return _file;
+    }
+}
+
++ (void) updateLogLevels {
+    CBLConsoleLogSink* console = self.console;
+    CBLLogLevel consoleLevel = console != nil ? console.level : kCBLLogLevelNone;
+    
+    CBLCustomLogSink* custom = self.custom;
+    CBLLogLevel customLevel = custom.logSink != nil ? custom.level : kCBLLogLevelNone;
+    
+    CBLFileLogSink* file = self.file;
+    CBLLogLevel fileLevel = file != nil ? file.level : kCBLLogLevelNone;
+    
+    CBLLogLevel callbackLevel = std::min(consoleLevel, customLevel);
+    CBLLogLevel domainsLevel = std::min(callbackLevel, fileLevel);
+    C4LogLevel c4Level = (C4LogLevel)domainsLevel;
+    
+    if (_domainsLevel != domainsLevel) {
+        for (NSString* domain in c4Domains) {
+            C4LogDomain c4domain = c4log_getDomain([domain UTF8String], false);
+            if (c4domain)
+                c4log_setLevel(c4domain, c4Level);
+        }
+        
+        for (NSString* domain in platformDomains) {
+            C4LogDomain c4domain = c4log_getDomain([domain UTF8String], false);
+            if (c4domain)
+                c4log_setLevel(c4domain, c4Level);
+        }
+        _domainsLevel = domainsLevel;
+    }
+    
+    if (_callbackLevel != callbackLevel) {
+        c4log_writeToCallback(c4Level, &c4Callback, true);
+        _callbackLevel = callbackLevel;
+    }
+}
+
+static void c4Callback(C4LogDomain c4domain, C4LogLevel c4level, const char *msg, va_list args) {
+    NSString* message = [NSString stringWithUTF8String: msg];
+    CBLLogLevel level = (CBLLogLevel) c4level;
+    CBLLogDomain domain = toCBLLogDomain(c4domain);
+    [CBLLogSinks.console writeLogWithLevel: level domain: domain message :message];
+    [CBLLogSinks.custom writeLogWithLevel: level domain: domain message :message];
+}
+
++ (void) writeCBLLog: (C4LogDomain)c4domain level: (C4LogLevel)c4level message: (NSString*)message {
+    CBLLogLevel level = (CBLLogLevel) c4level;
+    CBLLogDomain domain = toCBLLogDomain(c4domain);
+    
+    CBLStringBytes c4msg(message);
+    c4slog(c4domain, c4level, c4msg);
```

Comment:
When looking at this, I think CBLFileLogSink should implement CBLLogSinkProtocol protocol so this can be moved to CBLFileLogSink's writeLogWithLevel() the same way as the console and custom log sink below.